resolver/src/lookup_state.rs

```diff
@@ -54,6 +62,31 @@ impl<C: DnsHandle<Error = ResolveError> + 'static> CachingClient<C> {
     pub fn lookup(&mut self, query: Query) -> Box<Future<Item = Lookup, Error = ResolveError>> {
         QUERY_DEPTH.with(|c| *c.borrow_mut() += 1);
 
+        // see https://tools.ietf.org/html/rfc6761
+        //
+        // ```text
+        // Name resolution APIs and libraries SHOULD recognize localhost
+        // names as special and SHOULD always return the IP loopback address
+        // for address queries and negative responses for all other query
+        // types.  Name resolution APIs SHOULD NOT send queries for
+        // localhost names to their configured caching DNS server(s).
+        // ```
+        //
+        // special use rules only apply to the IN Class
+        if query.query_class() == DNSClass::IN {
+            match USAGE.get(query.name()).resolver() {
```

Comment:
I think this logic could be easily and efficiently implemented without using `USAGE` since the rules for localhost are very simple and `USAGE` is a quite heavyweight dependency. I think avoiding `USAGE` here in resolver would be ideal.

Reply:
Yes. Perhaps just .localhost., .local., and .test. Need to be in here for the Resolver.

I like the trie for the search primitives across all domains, but I think you’re generally against pulling that library in.